File: kotlin/src/main/kotlin/com/example/cote/LTCD/부분집합.kt
package com.example.cote.LTCD

import kotlin.random.*

val N = 15
val MAX_VALUE = 100
val TEST_CASE = 10

val random = Random(43)

fun dfs(nums: List<Int>, cur: Int, path: MutableList<Int>, ret: MutableList<List<Int>>) {
	
	ret.add(path.toList())

	for (i in cur..nums.size - 1) {
		path.add(nums[cur])
		dfs(nums, i + 1, path, ret)
		path.remove(nums[cur])
	}

}

fun solve(nums: List<Int>) : List<List<Int>> {
	val ret = mutableListOf<List<Int>>()
	val path = mutableListOf<Int>()

	dfs(nums, 0, path, ret)

	return ret
}

fun main() {
	
	repeat(TEST_CASE) {
		val n = random.nextInt(1, N)
		val nums = List(n) {
			random.nextInt(MAX_VALUE)
		}

		println(solve(nums))
	}
}

File: python/cote/LTCD/부분집합.py
from typing import List
import random

TEST_CASES = 10
MAX_VALUE = 100
N = 15

def solve(nums: List[int]) -> List[List[int]] :
	ret = []

	def dfs(i: int, path: List[int]) :
		ret.append(path)

		for j in range(i, len(nums)) :
			dfs(j + 1, [*path, nums[i]])

	dfs(0, [])

	return ret

if __name__ == "__main__" :

	for _ in range(TEST_CASES) :
		n = random.randrange(1,N)
		nums = random.sample(range(1,MAX_VALUE), n)
		print(solve(nums))
